Cache Book objects, use virtual files to avoid running out of fds.

hlinnaka opened the pull request:
Currently, whenever a page version is needed from an image or delta
layer, we open the file and read and parse the bookfile headers. That's
pretty expensive. To reduce the overhead, introduce a cache of open file
descriptors, and use that to cache the Book objects so that we don't need
to read the metadata on every access.

hlinnaka commented:
This makes a big difference e.g. to the 'vacuum' time in pgbench initialization. Before:
```
/tmp_install/bin/pgbench -s100 -i postgres
dropping old tables...
NOTICE:  table "pgbench_accounts" does not exist, skipping
NOTICE:  table "pgbench_branches" does not exist, skipping
NOTICE:  table "pgbench_history" does not exist, skipping
NOTICE:  table "pgbench_tellers" does not exist, skipping
creating tables...
generating data (client-side)...
10000000 of 10000000 tuples (100%) done (elapsed 6.97 s, remaining 0.00 s)
vacuuming...
creating primary keys...
done in 122.72 s (drop tables 0.00 s, create tables 0.01 s, client-side generate 12.09 s, vacuum 50.90 s, primary keys 59.71 s).
```
After:
```
...
done in 96.79 s (drop tables 0.00 s, create tables 0.01 s, client-side generate 11.52 s, vacuum 38.25 s, primary keys 47.02 s).
```

LizardWizzard commented on pageserver/src/virtual_file.rs:
```diff
@@ -0,0 +1,221 @@
+//! To track which files have been recently used, we the clock algorithm
+//! and a 'recently_used' flag on each slot.
+//!
+//! This is similar to PostgreSQL's vfd.c facility.
```
Probably a typo, I can only see a file named `fd.c`

LizardWizzard commented on pageserver/src/virtual_file.rs:
```diff
@@ -0,0 +1,221 @@
+    /// Lazy handle to the global file descriptor cache. The slot that this points to
+    /// might contain our File, or it may be empty, or it may contain a File that
+    /// belongs to a different VirtualFile.
+    handle: RwLock<SlotHandle>,
```
Could you please clarify why do we need `RwLock` here? In my understanding different threads can concurrently access one `Slot`, not `SlotHandle`. If we want to have multiple VirtualFile's pointing to the same SlotHandle it will require Rc/Arc. Please correct me if I'm wrong.

hlinnaka replied:
No, multiple VirtualFiles cannot have the same SlotHandle. But multiple threads can access the underlying file through a shared `&VirtualFile` reference, just like with `&File`. We need the RwLock so that if two threads try to read (or write) the file at the same time, using the same VirtualFile, only one thread will try to open the file and the other one will wait.

LizardWizzard replied:
> just like with &File

I see `impl Write for File` which contains `fn write(&mut self, buf: &[u8]) -> Result<usize>`. Note that it requires `&mut self`. So I do not see how it is possible to write to several &File concurrently from different threads directly. (Excluding into_raw_fd/from_raw_fd unsafe trickery and opening the same file from different threads).

I suggest to cover shared VirtualFiles with Arc/Mutex. So VirtualFile owns its SlotHandle, and if someone owns VIrtualFile there is no need to use locks.

hlinnaka replied:
See `FileExt::write_at`

knizhnik replied:
I wonder why do we need to implement non0reentrant Write trait for VirtualVile at all?  We can always use read_at/write_at...

hlinnaka replied:
True. Some code uses the Read/Write traits, also in external crates (e.g. serde), but we could provide an adapter struct that implements Read/Write on top of VirtualFile.

That said, it's also nice that VirtualFile is as close as possible to std::fs::File, so that it can be used as a drop-in replacement.

LizardWizzard replied:
> See FileExt::write_at

Oh, didn't see that. Good to know, thanks! 
Could you please add an example/test with concurrent access? So one can clearly see how it's possible with VirtualFile. Also do we have that case in our codebase (or where is it supposed to be used)? As far as I can see now we use only owned VirtualFile. So it looks like we are paying for flexibility we do not currently use. Please correct me if I misunderstood something.

knizhnik replied:
> 
> Hmm, you mean with actually accessing a file from two threads? That seems a bit complicated. I'm inclined to just trust the rust borrow checker here.

As far as FileExt methods are read-only, Rust doesn't prevent to call them from multiple threads.
I attach my test of multithreaded IO in Rust. May be you will want ti create test for VirtualFile from it.
[rust_io.zip](https://github.com/zenithdb/zenith/files/7513885/rust_io.zip)

LizardWizzard replied:
Sorry, I'm a little bit late, but nevertheless let it be here :)

> Hmm, you mean with actually accessing a file from two threads?

I think about it rather as a documentation. And as a smoke test too. Borrow checker cannot reason about rwlocks with different thread interleavings. Something like [loom](https://github.com/tokio-rs/loom) might reason about this better, but using it here looks like an overkill to me.

> BookFile uses the FileExt routines, with its read_chapter function

Yeah, but everywhere except this case with `inner` protected by a mutex we have an owned book nearby. That answers this question :) So we indeed need it for ~current code.

LizardWizzard commented on pageserver/src/virtual_file.rs:
```diff
@@ -0,0 +1,221 @@
+        // Upgrade lock and try again
+        let mut handle = self.handle.write().unwrap();
+        if handle.tag != INVALID_TAG {
+            let slot = OPEN_FILES.slots[handle.index].write().unwrap();
```
I wonder, is it possible to have a deadlock here?

hlinnaka replied:
No. This doesn't actually upgrade the lock (there is no function to do that in RwLock anyway), but it releases the old lock and re-acquires it in write mode.

There are two locks that are held simultaneously: the lock in VirtualFile, and the lock in Slot. The lock in VirtualFile is always acquired first.

LizardWizzard commented on pageserver/src/virtual_file.rs:
```diff
@@ -0,0 +1,221 @@
+        loop {
+            next = OPEN_FILES.next.fetch_add(1, Ordering::SeqCst) % OPEN_FILES.slots.len();
+
+            if retries < 100 {
```
May benefit from being a top level const

hlinnaka replied:
I changed this to be num_slots * 2.

LizardWizzard commented on pageserver/src/virtual_file.rs:
```diff
@@ -0,0 +1,221 @@
+
+impl Drop for VirtualFile {
+    fn drop(&mut self) {
+        // Upgrade lock and try again
```
bogus comment?

LizardWizzard reviewed (commented):
Nice! Could you please add unit tests for that?

knizhnik reviewed (commented):
You have told me in morning call that I do no see any problems in handling large number of files.
Well, there is 550 lines patch fixing this problem (222 lines in virtual_file). It seems to be not so large, but... all yakv implementation, including B-Tree, buffer pool, WAL, transactions is 2k lines. Just 4x times larger than this patch, which is actually playing in most popular "russian game": first create a problem from scratch, and then heroically fix it:)

And it is really non-trivial code which can suffer from starvation and may be even deadlock.
I once again wondering why you prefer lock to much simple LRU (as in Postgres fd.c?).
Yes, we have multithreaded access to the files in page server (unlike Postgres backend).
But it seems to me that pinning file descriptor under mutex can not be a bottleneck, taken in account that we are going to use it to read or write file.

Also, unlike Postgres fd.c there is no attempt to calculate actual number of available file descriptors for process (implementing which adds several more hundreds of lines of code). Hardcoded value 100 can easily cause deadlock if we have more than 100 active tenants and all of them will try to do something (i.e. execute get_page_at_lsn). I am not even mentioning that work with large number of tenants, having small databases, but with large number of relations (even system catalog contains several hundreds of relation) will be extremely inefficient: page server  will permanently open and close files.

hlinnaka commented:
> You have told me in morning call that you do no see any problems in handling large number of files. Well, there is 550 lines patch fixing this problem (222 lines in virtual_file). It seems to be not so large, but... all yakv implementation, including B-Tree, buffer pool, WAL, transactions is 2k lines. Just 4x times larger than this patch, which is actually playing in most popular "russian game": first create a problem from scratch, and then heroically fix it:)

I think we'll need a virtual file descriptor facility like this in any case. Even if we used one big file for each tenant, we would still want to support hundreds or thousands of tenants on a single page server, each of which would consume a file descriptor. And you would want to have plenty of headroom for other files and sockets.

> And it is really non-trivial code which can suffer from starvation and may be even deadlock. I once again wondering why you prefer clock to much simple LRU (as in Postgres fd.c?). Yes, we have multithreaded access to the files in page server (unlike Postgres backend). But it seems to me that pinning file descriptor under mutex can not be a bottleneck, taken in account that we are going to use it to read or write file.

Sure, an LRU list would work just as well.

> Also, unlike Postgres fd.c there is no attempt to calculate actual number of available file descriptors for process (implementing which will add several more hundreds of lines of code). Hardcoded value 100 can easily cause deadlock if we have more than 100 active tenants and all of them will try to do something (i.e. execute get_page_at_lsn). I am not even mentioning that work with large number of tenants, having small databases, but with large number of relations (even system catalog contains several hundreds of relation) will be extremely inefficient: page server will permanently open and close files.

Yeah, it would be nice to probe it at startup. Or just make it configurable and set it to right value; we control the infrastructure.

patins commented on pageserver/src/virtual_file.rs:
```diff
@@ -0,0 +1,221 @@
+            }
+        }
+
+        // Upgrade lock and try again
```
Why try again?

patins commented on pageserver/src/virtual_file.rs:
```diff
@@ -0,0 +1,221 @@
+        let mut retries = 0;
+        let mut slot;
+        let mut next;
+        loop {
+            next = OPEN_FILES.next.fetch_add(1, Ordering::SeqCst) % OPEN_FILES.slots.len();
+
+            if retries < 100 {
+                if let Ok(s) = OPEN_FILES.slots[next].try_write() {
+                    if !s.recently_used.load(Ordering::Relaxed) {
+                        slot = s;
+                        break;
+                    }
+                }
+                retries += 1;
+            } else {
+                slot = OPEN_FILES.slots[next].write().unwrap();
+                break;
+            }
+        }
```
Might be better written as a for loop.
What's the idea with using fetch_add? Is it so concurrent searchers don't try to grab the same slots? Do we need SeqCst?

hlinnaka replied:
> What's the idea with using fetch_add? Is it so concurrent searchers don't try to grab the same slots?

Yeah.

> Do we need SeqCst?

No, I think you're right, that's overkill. I changed it to AcqRel. An even weaker mode would probably work fine, but I'm not sure. This isn't a correctness issue in any case, even picking a slot in random would be OK, just not optimal.

patins commented:
Also we should add fsync to these APIs, maybe in a later PR

knizhnik commented:
> Sure, an LRU list would work just as well.
Actually I want to take my words back.
I remember that when I have implemented multithreaded posgres prototype (replace backend processes with threads),
my first attempt was to have shared cache of file descriptors (just put fd.c code in critical section).
But it immediately cause high contention and pgbench performance was several times smaller.
I have to use local file descriptor cache for each thread, although it definitely cause inefficient usage of open file descriptors. So may be using clock algorithm will help to solve this problem. But I afraid that try-lock loop is not he best thing to prevent conflicts.

knizhnik commented:
> I think we'll need a virtual file descriptor facility like this in any case. Even if we used one big file for each tenant, we would still want to support hundreds or thousands of tenants on a single page server, each of which would consume a file descriptor. And you would want to have plenty of headroom for other files and sockets.

I am not sure that we can handle more than 1000 active tenants at one page server. File descriptor is not the only per-tenant resource. There is also memory (buffer cache for KV or in-mem layers), threads (wal receiver, GC, checkpointer), socket ... So if we need to server large number of tenants, we need to implement some eviction policy for inactive tenants.
And it is much more than just closing file descriptor.

Also this virtual file implementation do not help to solve the problem with sockets. It is also requires file descriptor. It is really strange to have pool of opened files but doesn't take in account opened sockets, which number is almost the same.

Finally, we can have single KV storage for all tenants. If will completely solve the problem not only with file descriptors, but also with used memory.  But it cause some extra storage overhead by including tenant-if in key and require storage implementation to be very scalable, providing efficient concurrent access. Also common storage can do more efficient commits (like in Postgres: flush several commits using one fsync).

So from my point of view file descriptor cache is Postgres rudiment which we do not need. And shared storage is really an option which should be discussed.

hlinnaka commented:
On 09/11/2021 11:14, Konstantin Knizhnik wrote:
>     I think we'll need a virtual file descriptor facility like this in
>     any case. Even if we used one big file for each tenant, we would
>     still want to support hundreds or thousands of tenants on a single
>     page server, each of which would consume a file descriptor. And you
>     would want to have plenty of headroom for other files and sockets.
> 
> I am not sure that we can handle more than 1000 active tenants at one 
> page server. File descriptor is not the only per-tenant resource. There 
> is also memory (buffer cache for KV or in-mem layers), threads (wal 
> receiver, GC, checkpointer), socket ... So if we need to server large 
> number of tenants, we need to implement some eviction policy for 
> inactive tenants.
> And it is much more than just closing file descriptor.

Sure, need to do those things too.

> Also this virtual file implementation do not help to solve the problem 
> with sockets. It is also requires file descriptor. It is really strange 
> to have pool of opened files but doesn't take in account opened sockets, 
> which number is almost the same.

Yeah. Not much we can do about the sockets, but using fewer file 
descriptors for the files helps to keep enough fds available for socket.

> Finally, we can have single KV storage for all tenants. If will 
> completely solve the problem not only with file descriptors, but also 
> with used memory. But it cause some extra storage overhead by including 
> tenant-if in key and require storage implementation to be very scalable, 
> providing efficient concurrent access.

Big -1 on sharing the KV store between tenants. Dropping or moving a 
tenant from the page server should be fast. And if the data for one 
tenant gets corrupted for whatever reason, I'd like to isolate that to 
just that tenant. And there is no performance gain to be had from 
sharing a file between tenants. The only downside is using more file 
descriptors, which is solved by this PR.

Memory should be shared across tenants. And sure, there's a risk of 
isolation bugs there too. But let's not share the physical files.

> Also common storage can do more efficient commits (like in Postgres:
> flush several commits using one fsync).
fsyncing is not in the latency-critical path in the page server, so 
that's not relevant.

- Heikki

hlinnaka commented:
Updated the branch with a version 2. It's the same basic idea, but I added a ton of comments and refactored and made a lot of small changes all over the place. The size of the array (and hence the number of concurrent open files) is now also configurable. Please take another look.

patins commented on pageserver/src/layered_repository/delta_layer.rs:
```diff
@@ -139,6 +139,8 @@ pub struct DeltaLayerInner {
     /// loaded into memory yet.
     loaded: bool,
 
+    book: Option<Book<VirtualFile>>,
```
Cool that we're able to cache open books now

patins commented on pageserver/src/bin/pageserver.rs:
```diff
@@ -321,6 +330,12 @@ fn main() -> Result<()> {
+            Arg::with_name("max_file_descriptors")
+                .long("max_file_descriptors")
+                .takes_value(true)
+                .help("Max number of file descriptors to keep open for files"),
```
Does postgres do something for other "file descriptors" like sockets/stdio?
Might be good to make the scope of the virtual files more clear in `help`

hlinnaka commented on pageserver/src/virtual_file.rs:
```diff
@@ -0,0 +1,569 @@
+pub fn init(num_slots: usize) {
+    // safety: This assumes that init() is called exactly once, early in page
+    // server startup, when no concurrent threads are running yet.
+    unsafe { OPEN_FILES.init(num_slots) }
```
Done

knizhnik commented:
This test reproduces deadlock (at my notebook it requires about half minute):

[vfile.zip](https://github.com/zenithdb/zenith/files/7515804/vfile.zip)

hlinnaka commented:
On 10/11/2021 22:44, Konstantin Knizhnik wrote:
> This test reproduces deadlock (at my notebook it requires about half 
> minute):
> 
> vfile.zip <https://github.com/zenithdb/zenith/files/7515804/vfile.zip>

Works for me, I cannot reproduce it. Can you do a little bit of 
debugging with gdb to see how it's happening, please?

- Heikki

knizhnik commented:
{{{
(gdb) info threads
  Id   Target Id         Frame 
* 1    Thread 0x7f9efbaba7c0 (LWP 729) "vfile" 0x00007f9efb283d2d in __GI___pthread_timedjoin_ex (threadid=140320785200896, thread_return=0x0, abstime=0x0, 
  2    Thread 0x7f9efa8e7700 (LWP 730) "vfile" 0x00007f9efb287938 in futex_abstimed_wait (private=<optimized out>, abstime=0x0, expected=3, futex_word=0x7f9ef400101c)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:172
...
  92   Thread 0x7f9ec9dee700 (LWP 829) "vfile" 0x00007f9efb287938 in futex_abstimed_wait (private=<optimized out>, abstime=0x0, expected=3, futex_word=0x7f9ef400101c)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:172
(gdb) 
(gdb) thread 2
[Switching to thread 2 (Thread 0x7f9efa8e7700 (LWP 730))]
#0  0x00007f9efb287938 in futex_abstimed_wait (private=<optimized out>, abstime=0x0, expected=3, futex_word=0x7f9ef400101c)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:172
172	../sysdeps/unix/sysv/linux/futex-internal.h: No such file or directory.
(gdb) bt
#0  0x00007f9efb287938 in futex_abstimed_wait (private=<optimized out>, abstime=0x0, expected=3, futex_word=0x7f9ef400101c)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:172
#1  __pthread_rwlock_wrlock_full (abstime=0x0, rwlock=0x7f9ef4001010) at pthread_rwlock_common.c:706
#2  __GI___pthread_rwlock_wrlock (rwlock=0x7f9ef4001010) at pthread_rwlock_wrlock.c:27
#3  0x000055e35d6728c3 in std::sys::unix::rwlock::RWLock::write (self=0x7f9ef4001010)
    at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/sys/unix/rwlock.rs:75
#4  0x000055e35d6766bd in std::sys_common::rwlock::MovableRWLock::write (self=0x7f9ef4000ce8)
    at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/sys_common/rwlock.rs:100
#5  0x000055e35d66af65 in std::sync::rwlock::RwLock<T>::write (self=0x7f9ef4000ce8) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/sync/rwlock.rs:282
#6  0x000055e35d67738e in vfile::OpenFiles::find_victim_slot (self=0x55e35d8c8078 <vfile::OPEN_FILES+8>) at src/main.rs:124
#7  0x000055e35d66e719 in vfile::VirtualFile::with_file (self=0x7f9efa8e4ad0, func=...) at src/main.rs:250
#8  0x000055e35d677b60 in <vfile::VirtualFile as std::os::imp::unix::fs::FileExt>::read_at (self=0x7f9efa8e4ad0, buf=&mut [u8](size=8192) = {...}, offset=0)
    at src/main.rs:342
#9  0x000055e35d66dfa3 in std::os::imp::unix::fs::FileExt::read_exact_at (self=0x7f9efa8e4ad0, buf=&mut [u8](size=8192) = {...}, offset=0)
    at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/os/./unix/fs.rs:102
#10 0x000055e35d66ec64 in vfile::main::{{closure}} () at src/main.rs:391
#11 0x000055e35d66b839 in std::sys_common::backtrace::__rust_begin_short_backtrace (f=...)
    at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/sys_common/backtrace.rs:125
#12 0x000055e35d66ab16 in std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}} ()
    at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/thread/mod.rs:481
#13 0x000055e35d66c379 in <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once (self=..., _args=())
    at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/panic/unwind_safe.rs:271
#14 0x000055e35d67008c in std::panicking::try::do_call (data=0x7f9efa8e6cd8 "") at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/panicking.rs:403
#15 0x000055e35d6702cb in __rust_try () at library/std/src/panicking.rs:543
#16 0x000055e35d670001 in std::panicking::try (f=...) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/panicking.rs:367
#17 0x000055e35d66d649 in std::panic::catch_unwind (f=...) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/panic.rs:129
#18 0x000055e35d66a96a in std::thread::Builder::spawn_unchecked::{{closure}} () at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/thread/mod.rs:480
#19 0x000055e35d67360e in core::ops::function::FnOnce::call_once{{vtable-shim}} () at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/ops/function.rs:227
#20 0x000055e35d693a23 in <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once ()
    at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/alloc/src/boxed.rs:1636
#21 <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once () at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/alloc/src/boxed.rs:1636
#22 std::sys::unix::thread::Thread::new::thread_start () at library/std/src/sys/unix/thread.rs:106
#23 0x00007f9efb2826db in start_thread (arg=0x7f9efa8e7700) at pthread_create.c:463
#24 0x00007f9efaa09a3f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
}}}

knizhnik commented:
It is not actually deadlock and when I attached debugger to the process, then after some time all threads are terminated. But before ~20 minutes all of them were blocked and nothing happen (CPU usage was zero).

knizhnik commented:
Sometimes test successfully finished in 27 seconds, sometimes "hangs" after few seconds of work...

knizhnik commented:
Looks like hanging takes place at final stage when some threads are finished. For some reasons number of threads is always 92. Even if some of them are terminated (reported under gdb) still total number of threads is 92.

hlinnaka commented:
Thanks for the test case, @knizhnik! For some reason, that test program never hung on my laptop, but I was able to reproduce deadlocks when I modified it a little, so the same VirtualFile structs are shared by the concurrent threads. @LizardWizzard was right, we really should test that, after all. Opened a PR for that, including the test I used: https://github.com/zenithdb/zenith/pull/868

knizhnik commented:
Yeh,... still not working for me:(
